### ISSUE TITLE
Nextcloud non-default datadir results in empty configuration directory

### ISSUE THREAD

**Issue by wueestry** [NixOS/nixpkgs#356973]
## Describe the bug

Setting the data and the home directory to a custom directory results in an empty configuration directory at the target location and a "Configuration was not read or initialized correctly, not overwriting custom_location/config/config.php" error when accessing nextcloud via the browser.

## Steps To Reproduce

1. Enable nextcloud `services.nextcloud.enable = true;
2. Set `services.nextcloud.home` and `services.nextcloud.datadir` to a non-default path.
3. Try to launch nextcloud and/or see files in the defined directory if config.php file is present.

## Expected behavior

The same behaviour as with the default `/var/lib/nextcloud` with the config.php file being located in the config directory and therefore being able to login to the service.

## Screenshots

![2024-11-18_15:04:34](https://github.com/user-attachments/assets/cfa14936-fd5c-4c4e-9cca-feadba467a49)

## Additional context

My full configuration would be the following one. Perhaps I just have something completely wrong with the config itself.

```
    sops.secrets.nextcloud-admin = {
      owner = "nextcloud";
      group = "nextcloud";
    };

    services.nextcloud = {
      enable = true;
      configureRedis = true;
      package = pkgs.nextcloud29;
      hostName = "apollo.nextcloud.home";
      home = "/mnt/storage/nextcloud";
      datadir = "/mnt/storage/nextcloud";
      database.createLocally = true;
      extraAppsEnable = true;
      autoUpdateApps.enable = true;

      config = {
        dbtype = "pgsql";
        adminpassFile = config.sops.secrets.nextcloud-admin.path;
        adminuser = "admin";
      };
    };

    services.nginx.virtualHosts.${config.services.nextcloud.hostName}.listen = [
      {
        addr = "0.0.0.0";
        port = 8081;
      }
    ];
```

Only setting the datadir to the custom path results in the nextcloud folder being owned by `root:root` and results in the same behaviour with the config.php file not being created correctly.

## Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.59, NixOS, 24.05 (Uakari), 24.05.20241107.83fb6c0`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.8`
 - nixpkgs: `/nix/store/l2v78vdd33hvyyy33w9zih2ss60k2yms-source`

## Notify maintainers

@Ma27 @onny 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Ma27** [comment]
@wueestry my theory is that the file is created before the `/mnt/...` stuff gets mounted. Can you `umount` the datadir and check if there's a symlink in the mount's directory then?

How are you mounting the filesystem btw?

**Comment by wueestry** [comment]
What do you mean exactly? If I `umount` the datadir and then restart `nextcloud-setup.service` manually, there is nothing, just the empty mount folder. If I then rebuild the system the drive is then mounted automatically again.

This is how I currently mount the filesystem
```
  fileSystems."/mnt/storage" = {
    device = "/dev/disk/by-uuid/4ece4a2c-f2bb-4017-84e7-072233329253";
    fsType = "btrfs";
  };
```

I also have tried if adding
```
  systemd.services."nextcloud-setup" = {
    requires = ["mnt-storage.mount"];
    after = ["mnt-storage.mount"];
  };
```
could help solving the timing issue. But no luck.

**Comment by Ma27** [comment]
> What do you mean exactly

My current theory is that the `override.config.php` was written to `/mnt/storage/nextcloud/config` _before_ the filesystem got mounted.

> and then restart nextcloud-setup.service manually, there is nothing, just the empty mount folder

The file doesn't get created by `nextcloud-setup`, but earlier, by systemd-tmpfiles.

But given that this doesn't exist in there without a mount, this doesn't seem to be the case anyways.

Is there anything interesting in the logs from `systemd-tmpfiles`?
Also, can you reproduce the problem with a different home/datadir (that isn't `/var/lib/nextcloud`, but also not a mount-point)? Would be good to know whether or not this is related to the mount.

**Comment by wueestry** [comment]
Yes, I experience the same behaviour when using any other folder, which is not the default one.
How can I take a look at the systemd-tempfile logs for nextcloud?

**Comment by Ma27** [comment]
The unit should be `systemd-tmpfiles-setup`.

I just created a VM like this:

```nix
{
  nextcloud = { pkgs, ... }: {
    services.nextcloud = {
      enable = true;
      home = "/mnt/storage/nextcloud";
      datadir = "/mnt/storage/nextcloud";
      hostName = "nextcloud";
      https = false;
      database.createLocally = true;
      config = {
        adminpassFile = "${pkgs.writeText "adminpass" "foobar"}";
      };
    };
  };
}
```

and created it with `nixos-build-vms`.

As a result, I do get a correctly set up configuration directory within `/mnt/storage/nextcloud`.
Have you checked whether the other non-default testing directory you've used isn't a mountpoint?

**Comment by almino** [comment]
Is there any relation with my [issue](https://github.com/NixOS/nixpkgs/issues/403785)?

**Comment by ca5ua1** [comment]
> Is there any relation with my [issue](https://github.com/NixOS/nixpkgs/issues/403785)?

It seems they related. The workaround helps.

Issue still exist. And the workaround should be posted on wiki

_If you use datadir on another storage device (probably remote storage, e.g. NFS) and have problem with config.php not being updated - use this workaround_
```
    fileSystems."/var/lib/nextcloud" = {
      device = "/mnt/storage/nextcloud";
      options = [ "bind" ];
    };
```
Something like that

**Comment by Ma27** [comment]
Feel free to just add it to the wiki: wikis are intended to be edited by its users.
